src/game/LFG/LFGMgr.h

```diff
@@ -0,0 +1,138 @@
+/*
+ * This file is part of the CMaNGOS Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef MANGOSSERVER_LFGMGR_H
+#define MANGOSSERVER_LFGMGR_H
+
+#include <list>
+#include <map>
+
+#include "Policies/Singleton.h"
+#include "Common.h"
+
+enum ClassRoles
+{
+    LFG_ROLE_NONE   = 0x00,
+    LFG_ROLE_TANK   = 0x01,
+    LFG_ROLE_HEALER = 0x02,
+    LFG_ROLE_DPS    = 0x04
+};
+
+enum RolesPriority
+{
+    LFG_PRIORITY_NONE   = 0,
+    LFG_PRIORITY_LOW    = 1,
+    LFG_PRIORITY_NORMAL = 2,
+    LFG_PRIORITY_HIGH   = 3
+};
+
+enum PlayerLeaveMethod
+{
+    PLAYER_CLIENT_LEAVE = 0,
+    PLAYER_SYSTEM_LEAVE = 1
+};
+
+enum GroupLeaveMethod
+{
+    GROUP_CLIENT_LEAVE  = 0,
+    GROUP_SYSTEM_LEAVE  = 1
+};
+
+#define MAX_DPS_COUNT = 3
+
+struct LFGPlayerQueueInfo
+{
+    ClassRoles roleMask;
+    uint32 team;
+    uint32 areaId;
+    uint32 timeInLFG;
+    bool hasQueuePriority;
+    std::string name;
+    std::list<std::pair<ClassRoles, RolesPriority>> rolePriority;
+
+    void CalculateRoles(Classes playerClass);
+    void CalculateTalentRoles(Player* player);
+    RolesPriority GetRolePriority(ClassRoles role);
+};
+
+struct LFGGroupQueueInfo
+{
+    uint32 availableRoles;
+    uint32 dpsCount;
+    uint32 team;
+    uint32 areaId;
+    uint32 groupTimer;
+};
+
+struct MeetingStoneInfo
+{
+    uint32 area;
+    uint32 minlevel;
+    uint32 maxlevel;
+    char*  name;
+};
+
+typedef std::vector<MeetingStoneInfo> MeetingStoneSet;
+
+class LFGQueue
+{
+    public:
+        LFGQueue() {}
+        ~LFGQueue() {}
+
+        void AddToQueue(Player* leader, uint32 queAreaID);
+        void RestoreOfflinePlayer(Player* player);
+        bool IsPlayerInQueue(ObjectGuid const& plrGuid) const;
+        bool IsGroupInQueue(uint32 groupId) const;
+        void RemovePlayerFromQueue(ObjectGuid const& plrGuid, PlayerLeaveMethod leaveMethod = PLAYER_CLIENT_LEAVE); // 0 == by default system (cmsg, leader leave), 1 == by lfg system (no need report text you left queu)
```

Comment:
Using ObjectGuid as reference if its not being assigned to is overkill. ObjectGuid is a single uint64 value which is same as reference as we are using 64-bit systems, and with value the compiler can optimize the copies away.

Reply:
So it's never appropriate? I don't understand why they did it, but it seems it was on purpose, because there is this comment
https://github.com/cmangos/mangos-classic/pull/469/files#diff-f02ca7bf1bf633b9e81516bc6acc4c4c5275176077464cf070473fae85a5fb52R511

Do you have any ideas why could they do it? All I know this system supports working with offline players who got DC'ed so it keeps them in queue and doesn't look for replacement.

Reply:
Its fine to use ObjectGuid parameter event ObjectGuid&, but using ObjectGuid const& is overkill and unneeded, because if you cannot assign to it, there is no additional benefit.

Reply:
I'm sorry, how to fix it then? I see many similar things in code like this
https://github.com/cmangos/mangos-classic/search?p=1&q=ObjectGuid+const%26
Should I just use `bool IsPlayerInQueue(ObjectGuid plrGuid) const;` ?

Reply:
Then chill, we will change it during merge :) You will see whats what then.